pkg/controller/expand_and_recover.go

```diff
@@ -0,0 +1,234 @@
+		msg := fmt.Sprintf("Unable to expand %s because CSI driver %s only supports offline expansion and volume is currently in-use", util.PVCKey(pvc), ctrl.resizer.Name())
+		ctrl.eventRecorder.Event(pvc, v1.EventTypeWarning, util.VolumeResizeFailed, msg)
+		return pvc, pv, fmt.Errorf(msg), resizeNotCalled
+	}
+
+	// Record an event to indicate that external resizer is resizing this volume.
+	ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeResizing,
+		fmt.Sprintf("External resizer is resizing volume %s", pv.Name))
+
+	// before trying expansion we will remove the PVC from map
+	// that tracks PVCs which can't be expanded when in-use. If
+	// pvc indeed can not be expanded when in-use then it will be added
+	// back when expansion fails with in-use error.
+	ctrl.usedPVCs.removePVCWithInUseError(pvc)
+	pvc, pv, err = ctrl.callResizeOnPlugin(pvc, pv, newSize, pvcStatusSize)
+
+	if err != nil {
+		// Record an event to indicate that resize operation is failed.
+		ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeResizeFailed, err.Error())
+		return pvc, pv, err, true
+	}
+
+	klog.V(4).Infof("Update capacity of PV %q to %s succeeded", pv.Name, newSize.String())
+	return pvc, pv, nil, true
+}
+
+func (ctrl *resizeController) callResizeOnPlugin(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	newSize, oldSize resource.Quantity) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error) {
+	updatedSize, fsResizeRequired, err := ctrl.resizer.Resize(pv, newSize)
+
+	if err != nil {
+		// if this error was a in-use error then it must be tracked so as we don't retry without
+		// first verifying if volume is in-use
+		if inUseError(err) {
+			ctrl.usedPVCs.addPVCWithInUseError(pvc)
+		}
+		if isFinalError(err) {
+			var markExpansionFailedError error
+			pvc, markExpansionFailedError = ctrl.markControllerExpansionFailed(pvc)
+			if markExpansionFailedError != nil {
+				return pvc, pv, fmt.Errorf("resizing failed in controller with %v but failed to update PVC %s with: %v", err, util.PVCKey(pvc), markExpansionFailedError)
+			}
+		}
+		return pvc, pv, fmt.Errorf("resize volume %q by resizer %q failed: %v", pv.Name, ctrl.name, err)
+	}
+
+	klog.V(4).Infof("Resize volume succeeded for volume %q, start to update PV's capacity", pv.Name)
+
+	pv, err = ctrl.updatePVCapacity(pv, oldSize, updatedSize, fsResizeRequired)
+	if err != nil {
+		return pvc, pv, fmt.Errorf("error updating pv %q by resizer: %v", pv.Name, err)
+	}
+
+	if fsResizeRequired {
+		pvc, err = ctrl.markForPendingNodeExpansion(pvc)
+		return pvc, pv, err
+	}
+	pvc, err = ctrl.markOverallExpansionAsFinished(pvc, updatedSize)
+	return pvc, pv, err
+}
+
+// checks if pv can be expanded
+func (ctrl *resizeController) pvCanBeExpanded(pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) bool {
+	if !ctrl.resizer.CanSupport(pv, pvc) {
+		klog.V(4).Infof("Resizer %q doesn't support PV %q", ctrl.name, pv.Name)
```

Comment:
maybe it can be a warning because you are calling resize, but resizer does not support it?

Reply:
it can get called on all sort of PVs. such as intree pv,pvc etc, which are not resizable by external-resizer.  I think Info is suitable since most people will be running both intree expand_controller and this controller side-by-side for now.